tools/sgcloudspanner/emulator.go

```diff
@@ -15,17 +15,47 @@ import (
 	"go.einride.tech/sage/tools/sgdocker"
 )
 
+type Emulator struct {
+	ContainerID  string
+	DatabaseHost string
+	AdminHost    string
+	onClose      func()
+}
+
 // RunEmulator runs the Cloud Spanner emulator in Docker.
 func RunEmulator(ctx context.Context) (_ func(), err error) {
+	emulator, err := RunEmulatorContainer(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if emulator == nil {
+		return nil, nil
+	}
+	if err := os.Setenv("SPANNER_EMULATOR_HOST", emulator.DatabaseHost); err != nil {
+		emulator.Close()
+		return nil, err
+	}
+	cleanup := func() {
+		emulator.Close()
+		if err := os.Unsetenv("SPANNER_EMULATOR_HOST"); err != nil {
+			sg.Logger(ctx).Printf("failed to unset SPANNER_EMULATOR_HOST: %v", err)
+		}
+	}
+	return cleanup, nil
+}
+
+func RunEmulatorContainer(ctx context.Context) (emulator *Emulator, err error) {
```

Comment:
based on what the method is called, I would expect the return type to be `EmulatorContainer`?

Reply:
when the private state is removed, the return value becomes a pure value and we can remove the pointer from the return signature and return by value instead